ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/clustering/RCFSummarize.java

```diff
@@ -0,0 +1,161 @@
+        switch (distanceType) {
+            case L1:
+                distance = Summarizer::L1distance;
+                break;
+            case L2:
+                distance = Summarizer::L2distance;
+                break;
+            case LInfinity:
+                distance = Summarizer::LInfinitydistance;
+                break;
+            default:
+                distance = Summarizer::L2distance;
+                break;
+        }
+    }
+
+    @Override
+    public Model train(DataFrame dataFrame) {
+        Tuple<String[], float[][]> featureNamesValues = TribuoUtil.transformDataFrameFloat(dataFrame);
+        SampleSummary summary = Summarizer.summarize(featureNamesValues.v2(), 
+            parameters.getMaxK(), 
+            parameters.getInitialK(), 
+            parameters.getPhase1Reassign(), 
+            distance, 
+            new Random().nextLong(), 
+            parameters.getParallel());
+
+        Model model = new Model();
+        model.setName(FunctionName.RCF_SUMMARIZE.name());
+        model.setVersion(1);
+        model.setContent(ModelSerDeSer.serialize(new SerializableSummary(summary)));
+
+        return model;
+    }
+
+    @Override
+    public MLOutput predict(DataFrame dataFrame, Model model) {
+        if (model == null) {
+            throw new IllegalArgumentException("No model found for RCFSummarize prediction.");
+        }
+
+        SampleSummary summary = ((SerializableSummary)ModelSerDeSer.deserialize(model.getContent())).getSummary();
+        Iterable<float[]> centroidsLst = Arrays.asList(summary.summaryPoints);
+        Tuple<String[], float[][]> featureNamesValues = TribuoUtil.transformDataFrameFloat(dataFrame);
+        List<Integer> predictions = new ArrayList<>();
+        Arrays.stream(featureNamesValues.v2()).forEach(e->predictions.add(MathUtil.findNearest(e, centroidsLst, distance)));
+
+        List<Map<String, Object>> listClusterID = new ArrayList<>();
+        predictions.forEach(e -> listClusterID.add(Collections.singletonMap("ClusterID", e)));
+
+        return MLPredictionOutput.builder().predictionResult(DataFrameBuilder.load(listClusterID)).build();
+    }
+
+    @Override
+    public MLOutput trainAndPredict(DataFrame dataFrame) {
+        Tuple<String[], float[][]> featureNamesValues = TribuoUtil.transformDataFrameFloat(dataFrame);
+        SampleSummary summary = Summarizer.summarize(featureNamesValues.v2(), 
+            parameters.getMaxK(), 
+            parameters.getInitialK(), 
+            parameters.getPhase1Reassign(), 
+            distance, 
+            new Random().nextLong(), 
+            parameters.getParallel());
+        
+        Iterable<float[]> centroidsLst = Arrays.asList(summary.summaryPoints);
+        List<Integer> predictions = new ArrayList<>();
+        Arrays.stream(featureNamesValues.v2()).forEach(e->predictions.add(MathUtil.findNearest(e, centroidsLst, distance)));
```

Comment:
Just curious, I guess RCF has already calculated nearest centroid for all data points. Why doesn't RCF return the centroids for all data points directly (at least provide some option to let user decide whether output or not)? If so, we can avoid recalculating nearest centroids here.

Reply:
Let me dig into the code, I am not sure if the library supports exporting these things.

Reply:
Opps, they did export these things, but it is a good idea to follow.

Reply:
Thanks, I  created an issue to track this https://github.com/aws/random-cut-forest-by-aws/issues/338